### PR TITLE
Using multi bundle feature in AoP - Part 1

### PR DESCRIPTION
**Pull request by gaoran10**
### Motivation

Using multi bundle feature in AoP.

### Modifications

Supported operations
- Exchange Declare
- Queue Declare
- Queue Bind
- Queue Unbind
- Publish
- Consume

### Does this pull request potentially affect one of the following parts:

*If `yes` was chosen, please highlight the changes*

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API: (no)
  - The schema: (no)
  - The default values of configurations: (no)
  - The wire protocol: (no)
  - The rest endpoints: (no)
  - The admin cli options: (no)
  - Anything that affects deployment: (no)

### Documentation

Check the box below.

Need to update docs? 

- [ ] `doc-required` 
  
  (If you need help on updating docs, create a doc issue)
  
- [x] `no-need-doc` 
  
  (Please explain why)
  
- [ ] `doc` 
  
  (If this PR contains doc changes)

### CODE REVIEW

**Comment by mingyifei** [comment]
image

Missing stack information.

File: amqp-impl/src/main/java/io/streamnative/pulsar/handlers/amqp/AmqpMultiBundlesChannel.java [hunk 1]
```diff
@@ -0,0 +1,348 @@
...
+import org.apache.qpid.server.protocol.v0_8.transport.BasicConsumeOkBody;
+import org.apache.qpid.server.protocol.v0_8.transport.MessagePublishInfo;
+import org.apache.qpid.server.protocol.v0_8.transport.MethodRegistry;
+import org.apache.qpid.server.protocol.v0_8.transport.QueueDeclareOkBody;
+
+/**
+ * Amqp Channel level method processor.
+ */
+@Log4j2
+public class AmqpMultiBundlesChannel extends AmqpChannel {
```
Comment (mingyifei) [inline amqp-impl/src/main/java/io/streamnative/pulsar/handlers/amqp/AmqpMultiBundlesChannel.java:56]: The receiveBasicPublish method needs to be overridden to address default exchange issues such as:
```
public void receiveBasicPublish(AMQShortString exchange, AMQShortString routingKey, boolean mandatory, boolean immediate) {
        if (log.isDebugEnabled()) {
            log.debug("RECV[{}] BasicPublish[exchange: {} routingKey: {} mandatory: {} immediate: {}]", channelId, exchange, routingKey, mandatory, immediate);
        }
        if (isDefaultExchange(exchange)) {
            ExchangeDeclareParams params = new ExchangeDeclareParams();
            params.setType(ExchangeDefaults.DIRECT_EXCHANGE_CLASS);
            params.setInternal(false);
            params.setAutoDelete(false);
            params.setDurable(true);
            params.setPassive(false);
            params.setArguments(new HashMap<>());
            getAmqpAdmin().exchangeDeclare(connection.getNamespaceName().toString(), AbstractAmqpExchange.DEFAULT_EXCHANGE_DURABLE, params)
.thenCompose(__ -> {

                BindingParams bindingParams = new BindingParams();
                bindingParams.setRoutingKey(routingKey.toString());
                bindingParams.setArguments(new HashMap<>());

                return getAmqpAdmin().queueBind(connection.getNamespaceName().toString(), AbstractAmqpExchange.DEFAULT_EXCHANGE_DURABLE, routingKey.toString(), bindingParams).exceptionally(t -> {
                    log.error("Failed to bind queue {} to exchange {}.", routingKey, AbstractAmqpExchange.DEFAULT_EXCHANGE_DURABLE, t);
                    handleAoPException(t);
                    return null;
                });
            }).thenRun(() -> {
                MessagePublishInfo info = new MessagePublishInfo(AMQShortString.valueOf(AbstractAmqpExchange.DEFAULT_EXCHANGE_DURABLE), immediate, mandatory, routingKey);
                setPublishFrame(info, null);
            }).exceptionally(t -> {
                log.error("Failed to bind queue {} to exchange {}.", routingKey, AbstractAmqpExchange.DEFAULT_EXCHANGE_DURABLE, t);
                handleAoPException(t);
                return null;
            }).join();
        } else {
            MessagePublishInfo info = new MessagePublishInfo(exchange, immediate, mandatory, routingKey);
            setPublishFrame(info, null);
        }
    }
```

File: amqp-impl/src/main/java/io/streamnative/pulsar/handlers/amqp/ExchangeMessageRouter.java [hunk 2]
```diff
@@ -0,0 +1,451 @@
...
+        private final Map<Destination, TopicRoutingKeyParser> destinationMap;
+
+        public TopicExchangeMessageRouter(PersistentExchange exchange) {
+            super(exchange);
+            destinationMap = new ConcurrentHashMap<>();
+        }
+
+        static class TopicRoutingKeyParser {
+
+            Set<String> bindingKeys;
```
Comment (mingyifei) [inline amqp-impl/src/main/java/io/streamnative/pulsar/handlers/amqp/ExchangeMessageRouter.java:359]: The TopicRoutingKeyParser.bindingKeys needs to be initialized.

File: amqp-impl/src/main/java/io/streamnative/pulsar/handlers/amqp/AmqpMultiBundlesChannel.java [hunk 3]
```diff
@@ -0,0 +1,348 @@
...
+            return FutureUtil.failedFuture(e);
+        }
+        CompletableFuture<AmqpPulsarConsumer> consumerFuture = new CompletableFuture<>();
+        client.newConsumer()
+                .topic(getTopicName(PersistentQueue.TOPIC_PREFIX, queue))
+                .subscriptionType(SubscriptionType.Shared)
+                .subscriptionName("AMQP_DEFAULT")
+                .subscriptionInitialPosition(SubscriptionInitialPosition.Earliest)
+                .consumerName(UUID.randomUUID().toString())
+                .receiverQueueSize(getConnection().getAmqpConfig().getAmqpPulsarConsumerQueueSize())
```
Comment (mingyifei) [inline amqp-impl/src/main/java/io/streamnative/pulsar/handlers/amqp/AmqpMultiBundlesChannel.java:325]: When calling basicReject basicNack and requeue is true, use consumer.negativeAcknowledge API, its default value is 1 minute, should be manual configuration, such as: negativeAckRedeliveryDelay(0, TimeUnit.MILLISECONDS).

File: amqp-impl/src/main/java/io/streamnative/pulsar/handlers/amqp/ExchangeMessageRouter.java [hunk 4]
```diff
@@ -0,0 +1,451 @@
...
+            entry.release();
+            FutureUtil.waitForAll(futures).whenComplete((__, t) -> {
+                if (t != null) {
+                    log.error("Failed to route message {} for exchange {}.", position, exchange.exchangeName, t);
+                    cursor.rewind();
+                    tryToReadMoreEntries();
+                    return;
+                }
+
+                cursor.asyncDelete(position, new AsyncCallbacks.DeleteCallback() {
```
Comment (mingyifei) [inline amqp-impl/src/main/java/io/streamnative/pulsar/handlers/amqp/ExchangeMessageRouter.java:221]: Whether it is likely to be read again while being deleted.

File: amqp-impl/src/main/java/io/streamnative/pulsar/handlers/amqp/ExchangeMessageRouter.java [hunk 5]
```diff
@@ -0,0 +1,451 @@
...
+
+    private CompletableFuture<Void> sendMessage(MessageImpl<byte[]> msg, Destination des, int readerIndex) {
+        return getProducer(des.name, des.type)
+                .thenCompose(producer -> {
+                    msg.getMessageBuilder().clearProducerName();
+                    msg.getMessageBuilder().clearPublishTime();
+                    msg.getDataBuffer().readerIndex(readerIndex);
+                    return ((ProducerImpl<byte[]>) producer).sendAsync(msg);
+                })
+                .thenApply(__ -> null);
```
Comment (mingyifei) [inline amqp-impl/src/main/java/io/streamnative/pulsar/handlers/amqp/ExchangeMessageRouter.java:254]: No stack information when producer creation fails.

File: amqp-impl/src/main/java/io/streamnative/pulsar/handlers/amqp/admin/Queues.java [hunk 6]
```diff
@@ -77,13 +82,19 @@ public void getQueue(@Suspended final AsyncResponse response,
...
-                                QueueDeclareParams params) {
-        declareQueueAsync(vhost, queue, params)
+                             @PathParam("vhost") String vhost,
+                             @PathParam("queue") String queue,
+                             QueueDeclareParams params,
+                             @QueryParam("authoritative") @DefaultValue("false") boolean authoritative) {
+        TopicName topicName = TopicName.get(TopicDomain.persistent.toString(),
+                NamespaceName.get("public", vhost), PersistentQueue.TOPIC_PREFIX + queue);
+        validateTopicOwnershipAsync(topicName, authoritative)
+                .thenCompose(__ -> declareQueueAsync(vhost, queue, params))
```
Comment (mingyifei) [inline amqp-impl/src/main/java/io/streamnative/pulsar/handlers/amqp/admin/Queues.java:92]: I think we should pass the params all the way to the PersistentQueue.

File: amqp-impl/src/main/java/io/streamnative/pulsar/handlers/amqp/ExchangeMessageRouter.java [hunk 7]
```diff
@@ -0,0 +1,451 @@
...
+                        if (log.isDebugEnabled()) {
+                            log.debug("{} Deleted message at {}", exchange.getName(), ctx);
+                        }
+                    }
+
+                    @Override
+                    public void deleteFailed(ManagedLedgerException exception, Object ctx) {
+                        log.error("{} Failed to delete message at {}", exchange.getName(), ctx, exception);
+                    }
+                }, entry.getPosition());
```
Comment (mingyifei) [inline amqp-impl/src/main/java/io/streamnative/pulsar/handlers/amqp/ExchangeMessageRouter.java:233]: The entry has been released, ledgerId = -1,entryId = -1, entry.getPosition() is meaningless.

File: amqp-impl/src/main/java/io/streamnative/pulsar/handlers/amqp/ExchangeMessageRouter.java [hunk 8]
```diff
@@ -0,0 +1,451 @@
...
+
+    private void tryToReadMoreEntries() {
+        if (PENDING_SIZE_UPDATER.decrementAndGet(this) < replicatorQueueSize * 0.5
+                && HAVE_PENDING_READ_UPDATER.get(this) == FALSE) {
+            this.readMoreEntries();
+        }
+    }
+
+    private CompletableFuture<Void> sendMessage(MessageImpl<byte[]> msg, Destination des, int readerIndex) {
+        return getProducer(des.name, des.type)
```
Comment (mingyifei) [inline amqp-impl/src/main/java/io/streamnative/pulsar/handlers/amqp/ExchangeMessageRouter.java:247]: ByteBuf has retained(), we should release ByteBuf if we failed to create the producer or send the message.

File: amqp-impl/src/main/java/io/streamnative/pulsar/handlers/amqp/admin/impl/BaseResources.java [hunk 9]
```diff
@@ -19,30 +19,49 @@
...
+import org.apache.pulsar.metadata.api.MetadataStoreException;
 
 /**
  * Base resources.
  */
-
+@Slf4j
 public class BaseResources {
 
     protected String tenant = "public";
```
[the remark this answers is not in the document]
Comment (gaoran10) [reply amqp-impl/src/main/java/io/streamnative/pulsar/handlers/amqp/admin/impl/BaseResources.java:54]: fixed

File: amqp-impl/src/main/java/io/streamnative/pulsar/handlers/amqp/AmqpMultiBundlesChannel.java (hunk 1 above)
Comment (gaoran10) [reply amqp-impl/src/main/java/io/streamnative/pulsar/handlers/amqp/AmqpMultiBundlesChannel.java:56]: Yes, I want resolve this case in next PR, I add a TODO comment first.

File: amqp-impl/src/main/java/io/streamnative/pulsar/handlers/amqp/ExchangeMessageRouter.java (hunk 2 above)
Comment (gaoran10) [reply amqp-impl/src/main/java/io/streamnative/pulsar/handlers/amqp/ExchangeMessageRouter.java:359]: fixed

File: amqp-impl/src/main/java/io/streamnative/pulsar/handlers/amqp/AmqpMultiBundlesChannel.java (hunk 3 above)
Comment (gaoran10) [reply amqp-impl/src/main/java/io/streamnative/pulsar/handlers/amqp/AmqpMultiBundlesChannel.java:325]: Ok, I add TODO label first.

File: amqp-impl/src/main/java/io/streamnative/pulsar/handlers/amqp/ExchangeMessageRouter.java (hunk 5 above)
Comment (gaoran10) [reply amqp-impl/src/main/java/io/streamnative/pulsar/handlers/amqp/ExchangeMessageRouter.java:254]: Add error logs in method `getProducer(String des, String desType)` and remove failed future object from cache.

File: amqp-impl/src/main/java/io/streamnative/pulsar/handlers/amqp/ExchangeMessageRouter.java (hunk 7 above)
Comment (gaoran10) [reply amqp-impl/src/main/java/io/streamnative/pulsar/handlers/amqp/ExchangeMessageRouter.java:233]: fixed

File: amqp-impl/src/main/java/io/streamnative/pulsar/handlers/amqp/admin/Queues.java (hunk 6 above)
Comment (gaoran10) [reply amqp-impl/src/main/java/io/streamnative/pulsar/handlers/amqp/admin/Queues.java:92]: fixed

File: amqp-impl/src/main/java/io/streamnative/pulsar/handlers/amqp/ExchangeMessageRouter.java (hunk 8 above)
Comment (gaoran10) [reply amqp-impl/src/main/java/io/streamnative/pulsar/handlers/amqp/ExchangeMessageRouter.java]: I change to use sync method to create producer, please take a look again.

File: amqp-impl/src/main/java/io/streamnative/pulsar/handlers/amqp/admin/Exchanges.java [hunk 10]
```diff
@@ -80,12 +85,18 @@ public void getExchange(@Suspended final AsyncResponse response,
     public void declareExchange(@Suspended final AsyncResponse response,
                                 @PathParam("vhost") String vhost,
                                 @PathParam("exchange") String exchange,
-                                ExchangeDeclareParams params) {
-        declareExchange(vhost, exchange, params)
+                                ExchangeDeclareParams params,
+                                @QueryParam("authoritative") @DefaultValue("false") boolean authoritative) {
+        TopicName topicName = TopicName.get(TopicDomain.persistent.toString(),
+                NamespaceName.get("public", vhost), PersistentExchange.TOPIC_PREFIX + exchange);
```
Comment (mingyifei) [inline amqp-impl/src/main/java/io/streamnative/pulsar/handlers/amqp/admin/Exchanges.java:91]: NamespaceName.get("public", vhost), need to modify "public".

Comment (mingyifei) [reply amqp-impl/src/main/java/io/streamnative/pulsar/handlers/amqp/admin/Exchanges.java]: fixed

File: amqp-impl/src/main/java/io/streamnative/pulsar/handlers/amqp/AmqpMultiBundlesChannel.java (hunk 3 above)
Comment (mingyifei) [reply amqp-impl/src/main/java/io/streamnative/pulsar/handlers/amqp/AmqpMultiBundlesChannel.java:388]: fixed

File: amqp-impl/src/main/java/io/streamnative/pulsar/handlers/amqp/AmqpMultiBundlesChannel.java (hunk 1 above)
Comment (mingyifei) [reply amqp-impl/src/main/java/io/streamnative/pulsar/handlers/amqp/AmqpMultiBundlesChannel.java:55]: fixed

File: amqp-impl/src/main/java/io/streamnative/pulsar/handlers/amqp/ExchangeMessageRouter.java [hunk 11]
```diff
@@ -0,0 +1,474 @@
...
+                continue;
+            }
+
+            Set<Destination> destinations = getDestinations(
+                    props.getOrDefault(MessageConvertUtils.PROP_ROUTING_KEY, ""), getMessageHeaders());
+            initProducerIfNeeded(destinations);
+
+            final Position position = entry.getPosition();
+            List<CompletableFuture<MessageId>> futures = new ArrayList<>();
+            if (!destinations.isEmpty()) {
```
Comment (mingyifei) [inline amqp-impl/src/main/java/io/streamnative/pulsar/handlers/amqp/ExchangeMessageRouter.java:221]: if destinations.isEmpty, is it possible to ack exchange topic message?

File: amqp-impl/src/main/java/io/streamnative/pulsar/handlers/amqp/ExchangeMessageRouter.java [hunk 12]
```diff
@@ -0,0 +1,474 @@
...
+                                            Map<String, Object> arguments) {
+            messageRouterMap.computeIfAbsent(new Destination(des, desType), k -> new HeadersMessageRouter())
+                    .getArguments().putAll(arguments);
+        }
+
+        @Override
+        public synchronized void removeBinding(String des, String desType, String routingKey,
+                                               Map<String, Object> arguments) {
+            messageRouterMap.computeIfPresent(new Destination(des, desType), (k, v) -> {
+                v.getArguments().putAll(arguments);
```
Comment (mingyifei) [inline amqp-impl/src/main/java/io/streamnative/pulsar/handlers/amqp/ExchangeMessageRouter.java:454]: need to delete

File: amqp-impl/src/main/java/io/streamnative/pulsar/handlers/amqp/ExchangeServiceImpl.java [hunk 13]
```diff
@@ -189,4 +189,17 @@ public CompletableFuture<Integer> exchangeBound(NamespaceName namespaceName, Str
         return future;
     }
 
+    @Override
+    public CompletableFuture<Void> queueBind(NamespaceName namespaceName, String exchange, String queue,
+                                             String routingKey, Map<String, Object> arguments) {
+        return exchangeContainer.asyncGetExchange(namespaceName, exchange, false, null)
```
Comment (mingyifei) [inline amqp-impl/src/main/java/io/streamnative/pulsar/handlers/amqp/ExchangeServiceImpl.java:195]: If exchange is the system default (amp.topic, amqp.direct... ..) , should be created before bind

File: amqp-impl/src/main/java/io/streamnative/pulsar/handlers/amqp/ExchangeMessageRouter.java (hunk 11 above)
Comment (gaoran10) [reply amqp-impl/src/main/java/io/streamnative/pulsar/handlers/amqp/ExchangeMessageRouter.java:221]: Yes, if the destinations is empty, the futures will be also empty.

File: amqp-impl/src/main/java/io/streamnative/pulsar/handlers/amqp/ExchangeMessageRouter.java (hunk 12 above)
Comment (gaoran10) [reply amqp-impl/src/main/java/io/streamnative/pulsar/handlers/amqp/ExchangeMessageRouter.java:454]: The `arguments` is null while removing binding. I change this to remove the destination directly.

File: amqp-impl/src/main/java/io/streamnative/pulsar/handlers/amqp/ExchangeServiceImpl.java (hunk 13 above)
Comment (gaoran10) [reply amqp-impl/src/main/java/io/streamnative/pulsar/handlers/amqp/ExchangeServiceImpl.java:195]: I move the build-in exchange name check to method `ExchangeContainer#asyncGetExchange`.

**Review by mingyifei — APPROVED** [review]

**Review by codelipenghui — APPROVED** [review]

**Pull request merged**